<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Johannes Gawron" />

<meta name="date" content="2024-06-07" />

<title>Br38</title>

<script src="Br38_files/header-attrs-2.26/header-attrs.js"></script>
<script src="Br38_files/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="Br38_files/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="Br38_files/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="Br38_files/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="Br38_files/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="Br38_files/navigation-1.1/tabsets.js"></script>
<link href="Br38_files/highlightjs-9.12.0/default.css" rel="stylesheet" />
<script src="Br38_files/highlightjs-9.12.0/highlight.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>



<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div id="header">



<h1 class="title toc-ignore">Br38</h1>
<h4 class="author">Johannes Gawron</h4>
<h4 class="date">2024-06-07</h4>

</div>


<div id="splitting-statistics" class="section level2">
<h2>Splitting statistics</h2>
<p>This code analyses splitting statistics for CTC-clusters.</p>
<p>The analysis takes a list of trees sampled from its posterior
distribution as input and computes the mutations placement probability
distribution for each oneof them. From this distribution we derive a
score that quantifies the probability that two cells have experienced
divergent evolution. This score is called the splitting score.</p>
</div>
<div id="configure-the-script" class="section level2">
<h2>Configure the script</h2>
<pre class="r"><code>inputFolder &lt;- &quot;/Users/jgawron/Documents/projects/CTC_backup/input_folder&quot;
simulationInputFolder &lt;- &quot;/Users/jgawron/Documents/projects/CTC_backup/simulations/simulations2&quot;
treeName &lt;- &quot;Br38&quot;
nTreeSamplingEvents &lt;- 1000
nMutationSamplingEvents &lt;- 1000</code></pre>
</div>
<div id="loading-data" class="section level2">
<h2>Loading data</h2>
<pre class="r"><code>source(&quot;/Users/jgawron/Documents/projects/CTC-SCITE/CTC-SCITE/experiments/workflow/resources/functions.R&quot;)</code></pre>
<pre><code>## ── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
## ✔ dplyr     1.1.4     ✔ readr     2.1.5
## ✔ forcats   1.0.0     ✔ stringr   1.5.1
## ✔ ggplot2   3.5.1     ✔ tibble    3.2.1
## ✔ lubridate 1.9.3     ✔ tidyr     1.3.1
## ✔ purrr     1.0.2     
## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()
## ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
<pre class="r"><code>input &lt;- load_data(inputFolder, treeName)</code></pre>
<pre><code>## Rows: 40000 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (1): Tree
## dbl (4): LogScore, SequencingErrorRate, DropoutRate, LogTau
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 44 Columns: 26
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (3): X1, X3, X4
## dbl (23): X2, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X1...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 11 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (2): Cluster, Description
## dbl (3): CellCount, TCs, WBCs
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
</div>
<div id="sample-description" class="section level2">
<h2>Sample description</h2>
<p>Each row corresponds to a cell. Column description: - Cluster: An
number indicating the sample the cell belongs to. - ClusterName: The
name of the sample in the nodeDescription.tsv file - WBC: a binary
vector indicating whether the cell is a white blood cell (1) or not (0).
- color: Indicates the color of the cluster in the tree, as described in
the nodeDescription.tsv file.</p>
<pre class="r"><code>print(input$sample_description)</code></pre>
<pre><code>##    Cluster ClusterName WBC      color single_cell
## 1        0      Br38_1   0     gray93        TRUE
## 2        1     Br38_10   0     gray93        TRUE
## 3        2     Br38_11   0     gray93        TRUE
## 4        3      Br38_2   0     gray93        TRUE
## 5        4      Br38_3   0     gray93        TRUE
## 6        5      Br38_4   0     gray93        TRUE
## 7        6      Br38_5   0     gray93        TRUE
## 8        7      Br38_6   0     gray93        TRUE
## 9        8      Br38_7   0     gray93        TRUE
## 10       9      Br38_8   0 lightcoral       FALSE
## 11       9      Br38_8   0 lightcoral       FALSE
## 12       9      Br38_8   0 lightcoral       FALSE
## 13      10      Br38_9   0     gray93        TRUE</code></pre>
</div>
<div id="general-overview" class="section level2">
<h2>General overview</h2>
<p>We sample 1000 trees.</p>
<p>For each pair of cells in the same cluster and each sampled tree we
compute the splitting score, that is, the probability that the two cells
have experienced divergent evolution. A low splitting score (close to 0)
indicates that the two cells are likely genealogically closely related,
while a high splitting score (close to 1) indicates that the two cells
have evolved in a divergent manner.</p>
<p>Throughout the sampling of trees, this gives rise to an empirical
distribution of splitting scores for each pair of cells in the same
cluster. Intuitively, this distribution takes into account the
uncertainty in the tree estimation. To be able to interpret the
splitting score appropriately (e.g. to answer the question when is a
splitting score is high enough to call oligo-clonality) we need to
calibrate our expectations.</p>
<p>We do this by assessing the distributions of splitting scores when we
know that the clusters is mono-clonal. To this end, we simulate
reference and alternative read count data of monoclonal clusters of
different sizes (2,3,4 and 5-cell clusters) and add these to the
original dataset, run the tree inference algorithm and compute the
splitting score distributions for all pairs of cells in the same
cluster. To ensure that the simulated data does not confound the tree
inference too much, we do this one cluster at a time.</p>
<p>For each simulated cluster we pick one pair of cells and printed the
splitting score distribution below. With high number of sampled trees,
the distributions of all pairs of cells from the same cluster are very
similar, since the model treats all cells from the same cluster as
interchangeable.</p>
<p>Finally, we print the empirical distribution of the the splitting
scores for all clusters of the same size.</p>
<p>The latter is used to specify the cutoff for oligo-clonality: It is
defined as the 95%-percentile of the aggregated distribution of
splitting scores.</p>
<pre class="r"><code>cutoffsSplittingProbs &lt;- data.frame(clusterSize = vector(), Cutoff = vector())
cutoffsBranchingProbabilities &lt;- data.frame(clusterSize = vector(), Cutoff = vector())

for (clusterSize in 2:5){
  try(
  {treeNameSimulated &lt;- paste(treeName, clusterSize, sep = &#39;_&#39;)


  inputSimulated &lt;- load_data(simulationInputFolder, treeNameSimulated)

  sampleDescriptionSimulated &lt;- inputSimulated$sample_description
  
  distance &lt;- computeClusterSplits(inputSimulated$sample_description, inputSimulated$postSampling, treeNameSimulated, inputSimulated$nCells,
                     inputSimulated$nMutations, inputSimulated$nClusters,
                     inputSimulated$alleleCount,
                     inputSimulated$mutatedReadCounts, inputSimulated$totalReadCounts,
                     nMutationSamplingEvents = nMutationSamplingEvents, nTreeSamplingEvents = nTreeSamplingEvents,
                     cellPairSelection = c(&quot;orchid&quot;, &quot;orchid1&quot;, &quot;orchid2&quot;,
                                           &quot;orchid3&quot;, &quot;orchid4&quot;, &quot;darkorchid&quot;,
                                           &quot;darkorchid1&quot;,&quot;darkorchid2&quot;, &quot;darkorchid3&quot;,
                                           &quot;darkorchid4&quot;, &quot;purple&quot;, &quot;purple1&quot;,
                                           &quot;purple2&quot;, &quot;purple3&quot;, &quot;purple4&quot;))

  
  cutoffsSplittingProbs &lt;- rbind(cutoffsSplittingProbs, data.frame(clusterSize = clusterSize, Cutoff = mean(distance$splittingProbs$Splitting_probability) + 2 * sd(distance$splittingProbs$Splitting_probability) ))

  plot(ggplot(data.frame(x = distance$aggregatedBranchingProbabilities), aes(x = x)) +
    geom_histogram(binwidth = 0.01))
  cutoffsBranchingProbabilities &lt;- rbind(cutoffsBranchingProbabilities, data.frame(clusterSize = clusterSize, Cutoff = quantile(distance$aggregatedBranchingProbabilities, probs = 0.95, names = FALSE)[1] ))
  })
}</code></pre>
<pre><code>## Rows: 1136 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (1): Tree
## dbl (4): LogScore, SequencingErrorRate, DropoutRate, LogTau
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 44 Columns: 34
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (3): X1, X3, X4
## dbl (31): X2, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X1...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 15 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (2): Cluster, Description
## dbl (3): CellCount, TCs, WBCs
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p><img src="Br38_files/figure-html/computing-simulated-clusters-1.png" width="672" /><img src="Br38_files/figure-html/computing-simulated-clusters-2.png" width="672" /><img src="Br38_files/figure-html/computing-simulated-clusters-3.png" width="672" /><img src="Br38_files/figure-html/computing-simulated-clusters-4.png" width="672" /><img src="Br38_files/figure-html/computing-simulated-clusters-5.png" width="672" /></p>
<pre><code>## Rows: 255 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (1): Tree
## dbl (4): LogScore, SequencingErrorRate, DropoutRate, LogTau
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 44 Columns: 32
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (3): X1, X3, X4
## dbl (29): X2, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X1...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 14 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (2): Cluster, Description
## dbl (3): CellCount, TCs, WBCs
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 105 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (1): Tree
## dbl (4): LogScore, SequencingErrorRate, DropoutRate, LogTau
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 44 Columns: 30
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (3): X1, X3, X4
## dbl (27): X2, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X1...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 13 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (2): Cluster, Description
## dbl (3): CellCount, TCs, WBCs
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 145 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (1): Tree
## dbl (4): LogScore, SequencingErrorRate, DropoutRate, LogTau
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 44 Columns: 30
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr  (3): X1, X3, X4
## dbl (27): X2, X5, X6, X7, X8, X9, X10, X11, X12, X13, X14, X15, X16, X17, X1...
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
## Rows: 13 Columns: 5
## ── Column specification ────────────────────────────────────────────────────────
## Delimiter: &quot;\t&quot;
## chr (2): Cluster, Description
## dbl (3): CellCount, TCs, WBCs
## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<p><img src="Br38_files/figure-html/computing-simulated-clusters-6.png" width="672" /></p>
<pre class="r"><code>print(cutoffsBranchingProbabilities)</code></pre>
<p>Now we can compute the aggregated splitting score distributions for
each cluster. The distribution’s mean is compared to the cutoffs
computed above, and if it is higher than the cutoff, we call the cluster
oligo-clonal.</p>
<pre class="r"><code>nTumorClusters &lt;- 0
nOligoclonalClusters2 &lt;- 0
splittingSummary2 &lt;- data.frame(Color = vector(), Oligoclonal = vector(), ClusterSize = vector())

for(clusterSize in 2:5){
  try({
    clusterColor &lt;- input$sample_description %&gt;%
    filter(WBC ==0 &amp;  color != &#39;gray93&#39;) %&gt;%
    group_by(color) %&gt;%
    filter(n() == clusterSize) %&gt;%
    pull(color) %&gt;%
    unique() 
    
    for(color in clusterColor){
      distance &lt;- computeClusterSplits(input$sample_description, input$postSampling, treeName, input$nCells,
                     input$nMutations, input$nClusters,
                     input$alleleCount,
                     input$mutatedReadCounts, input$totalReadCounts,
                     nMutationSamplingEvents = nMutationSamplingEvents, nTreeSamplingEvents = nTreeSamplingEvents,
                     cellPairSelection = c(color))

      splittingProbs &lt;- mean(distance$splittingProbs$Splitting_probability)
      branchingProbs &lt;- mean(distance$aggregatedBranchingProbabilities)
      
      nTumorClusters &lt;- nTumorClusters + 1
      oligoclonal &lt;- FALSE

      if(branchingProbs &gt; cutoffsBranchingProbabilities[(cutoffsBranchingProbabilities$clusterSize == clusterSize), 2]){
        nOligoclonalClusters2 &lt;- nOligoclonalClusters2 + 1
        oligoclonal &lt;- TRUE
      }
      splittingSummary2 &lt;- rbind(splittingSummary2, data.frame(Color = color, Oligoclonal = oligoclonal, ClusterSize = clusterSize))
    }
  })
}</code></pre>
<pre><code>## [1] &quot;Computing genomic distances of leaves: 10 9&quot;
## [1] &quot;Computing the posterior distribution&quot;</code></pre>
<p><img src="Br38_files/figure-html/computing-real-clusters-1.png" width="672" /><img src="Br38_files/figure-html/computing-real-clusters-2.png" width="672" /></p>
<pre><code>## Error in if (branchingProbs &gt; cutoffsBranchingProbabilities[(cutoffsBranchingProbabilities$clusterSize ==  : 
##   Argument hat Länge 0</code></pre>
<pre class="r"><code>numberOfCancerClusters &lt;- input$sample_description %&gt;%
    filter(WBC ==0 &amp;  color != &#39;gray93&#39;) %&gt;%
    group_by(color) %&gt;%
    filter(n() &gt; 1) %&gt;%
    pull(color) %&gt;%
    unique() %&gt;% length() 

print(sprintf(&#39;%d out of %d clusters were found to be oligoclonal in %s, using method 2&#39;, nOligoclonalClusters2, numberOfCancerClusters, treeName))</code></pre>
<pre><code>## [1] &quot;0 out of 1 clusters were found to be oligoclonal in Br38, using method 2&quot;</code></pre>
<pre class="r"><code>print(splittingSummary2)</code></pre>
<pre><code>## [1] Color       Oligoclonal ClusterSize
## &lt;0 Zeilen&gt; (oder row.names mit Länge 0)</code></pre>
</div>




</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
