---
title: "Br7"
author: "Johannes Gawron"
date: "2024-02-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Splitting statistics

This code analyses splitting statistics for CTC-clusters. The analysis takes a list of trees sampled from its posterior distribution as input.

and samples mutations placements for each of the trees.


## Configure the script
```{r config}
inputFolder <- "../../input_folder"
treeName <- "Br7"
```

## Loading data
```{r load}
source("functions.R")

input <- load_data(inputFolder, treeName)

postSampling <- input$postSampling
nClusters <- input$nClusters
ClusterID <- input$clusterID
nCells <- input$nCells  
nMutations <- input$nMutations
nClusters <- input$nClusters
alleleCount <- input$alleleCount
mutatedReadCounts <- input$mutatedReadCounts
totalReadCounts <- input$totalReadCounts
sampleDescription <- input$sample_description
mutationDescription <- input$mutationDescription
annotations <- input$annotations
```



## Sample description 

Each row corresponds to a cell.
Column description:
  - Cluster: An number indicating which sample the cell belongs to.
  - ClusterName: The name of the sample in the nodeDescription.tsv file
  - WBC: a binary vector indicating whether the cell is a white blood cell (1) or not (0).
  - color: Indicates the color of the cluster in the tree, as described in the nodeDescription.tsv
          file.
  
```{r Describe samples}
print(sampleDescription)
```

## Null distributions

To get a grasp of the null distribution of cluster splits, we will have a look at 
  - the splitting statistics for single cells
  - simulated CTC clusters that contain the aggregated read counts of the single cells
  where these single cells have the same or a similar genotype according to the mutation
  call using the genotype calls from Monovar.

```{r identify candidates for monoclonal pairs}
candidate_pairs <- load_monoclonal_pairs(inputFolder, treeName)
print(candidate_pairs$monoclonal_pairs)
print(candidate_pairs$distance_matrix)
```

By visual inspection of the pairwise distance matrix of genotypes (only the lower triangular matrix computed)
only genetically very distinct single cells could be identified. So I will skip
this step for Br7 and go to the splitting statistics right away.


## Splitting statistics
Now I am going through all clusters and comparing all tumour cells within a cluster
with each other.
This is done by
  - going through all sampled tree
  - for each of the trees sample mutation placements from the posterior distribution
  - for each tree with mutation placement compute the evolutionary distance of the two 
    cells
  - This will result in a histogram of evolutionary distances for each pair of
    sampled cells. We report the median of this histogram.
    
Two parameters can be tuned: The number of sampled trees and the number of mutation
placements sampled for each of the trees.

```{r Compute the splitting statistics}

nTreeSamplingEvents <- 100
nMutationSamplingEvents <- 100

distance <- computeClusterSplits(sampleDescription, postSampling, treeName, nCells,
                     nMutations, nClusters,
                     alleleCount,
                     mutatedReadCounts, totalReadCounts,
                     nMutationSamplingEvents = nMutationSamplingEvents, nTreeSamplingEvents = nTreeSamplingEvents)
```





In this specific case, there is genetic evidence indicating that one of the the WBCs is actually a tumour cell:


```{r WBC or tumor}
print(paste(paste("Computing genomic distances of leaves:", 20, sep = " "), 21, sep = " "))


desired_values <- sample(1:length(postSampling), size = nTreeSamplingEvents, replace = FALSE) %>% sort()
postSampling2 <- postSampling[desired_values]
  
distance <- c(distance, produce_Distance_Posterior(20,21,postSampling2, treeName, nCells,
                                                         nMutations, nClusters,
                                                         alleleCount,ClusterID,
                                                           mutatedReadCounts,totalReadCounts,sampleDescription$WBC,
                                                         nSamplingEvents = nMutationSamplingEvents, clusterName = "lemonchiffon")) 

```



## Filtering driver mutations

We classified mutations into driver mutations and mutations we are agnostic about.

```{r filter out non-drivers}
mutationFilter <- apply(mutationDescription, 1, FUN = IsDriver, annotations)
nMutations <- mutationFilter %>% sum() 
mutatedReadCounts <- mutatedReadCounts[mutationFilter]
totalReadCounts <- totalReadCounts[mutationFilter]
print(nMutations)
```

None of the mutations were annotated as oncogenic or driver.