---
title: "Pr9"
author: "Johannes Gawron"
date: "2024-02-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Splitting statistics

This code analyses splitting statistics for CTC-clusters. The analysis takes a list of trees sampled from its posterior distribution as input.

and samples mutations placements for each of the trees.


## Configure the script
```{r config}
inputFolder <- "../../input_folder"
treeName <- "Pr9"
```

## Loading data
```{r load}
source("functions.R")


input <- load_data(inputFolder, treeName)

postSampling <- input$postSampling
nClusters <- input$nClusters
ClusterID <- input$clusterID
nCells <- input$nCells  
nMutations <- input$nMutations
nClusters <- input$nClusters
alleleCount <- input$alleleCount
mutatedReadCounts <- input$mutatedReadCounts
totalReadCounts <- input$totalReadCounts
sampleDescription <- input$sample_description
```



## Sample description 

Each row corresponds to a cell.
Column description:
  - Cluster: An number indicating which sample the cell belongs to.
  - ClusterName: The name of the sample in the nodeDescription.tsv file
  - WBC: a binary vector indicating whether the cell is a white blood cell (1) or not (0).
  - color: Indicates the color of the cluster in the tree, as described in the nodeDescription.tsv
          file.
  
```{r Describe samples}
print(sampleDescription)
```

## Null distributions

To get a grasp of the null distribution of cluster splits, we will have a look at 
  - the splitting statistics for single cells
  - simulated CTC clusters that contain the aggregated read counts of the single cells
  where these single cells have the same or a similar genotype according to the mutation
  call using the genotype calls from Monovar.
Similar means that their genetic distance lies in the 1% quantile of the set of all pairwise genetic distances.
The Hamming distance is chosen.

```{r identify candidates for monoclonal pairs}
candidate_pairs <- load_monoclonal_pairs(inputFolder, treeName)
print(candidate_pairs$monoclonal_pairs)
print(candidate_pairs$distance_matrix)
```

Some of the candidate cells were filtered out before tree reconstruction, and others
are no tumor cells. Candidates are picked manually.

```{r identify genetically identical cells}
monoclonal_pairs <- list(c("Pr9_CTC_23","Pr9_CTC_13"), c("Pr9_CTC_24","Pr9_CTC_15"), c("Pr9_CTC_24", "Pr9_CTC_18"), c("Pr9_CTC_9", "Pr9_CTC_20"))
```

The triplet "Pr9_CTC_24", "Pr9_CTC_15", "Pr9_CTC_18" might be a candidate for 
simulating a 3-cell CTC cluster, but 15-18 was not suggested the output. Maybe these
two are still genetically close enough to count then as monoclonal:


```{r identify monoclonal triplet}
candidate_pairs$full_distance_matrix[c("Pr9_CTC_15.realigned.bam", "Pr9_CTC_18.realigned.bam","Pr9_CTC_24.realigned.bam"),c("Pr9_CTC_15.realigned.bam", "Pr9_CTC_18.realigned.bam","Pr9_CTC_24.realigned.bam")]
```

The genetic distance between "Pr9_CTC_15" and "Pr9_CTC_18" is 6 which is just above the 1% quantile, so I'll accept it.

This computes baseline distance histograms for single cells in the tree.

```{r baseline distance distributions}

nTreeSamplingEvents <- 100
nMutationSamplingEvents <- 100


distance <- computeClusterSplits(sampleDescription, postSampling, treeName, nCells,
                     nMutations, nClusters,
                     alleleCount,
                     mutatedReadCounts, totalReadCounts,
                     nMutationSamplingEvents = nMutationSamplingEvents, nTreeSamplingEvents = nTreeSamplingEvents, cellPairSelection = monoclonal_pairs)

```



## Splitting statistics
Now I am going through all clusters and comparing all tumour cells within a cluster
with each other.
This is done by
  - going through all sampled tree
  - for each of the trees sample mutation placements from the posterior distribution
  - for each tree with mutation placement compute the evolutionary distance of the two 
    cells
  - This will result in a histogram of evolutionary distances for each pair of
    sampled cells. We report the median of this histogram.
    
Two parameters can be tuned: The number of sampled trees and the number of mutation
placements sampled for each of the trees.

```{r Compute the splitting statistics}


distance <- computeClusterSplits(sampleDescription, postSampling, treeName, nCells,
                     nMutations, nClusters,
                     alleleCount,
                     mutatedReadCounts, totalReadCounts,
                     nMutationSamplingEvents = nMutationSamplingEvents, nTreeSamplingEvents = nTreeSamplingEvents)
```


